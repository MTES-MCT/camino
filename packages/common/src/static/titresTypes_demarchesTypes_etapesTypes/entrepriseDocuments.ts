import { isNotNullNorUndefined, onlyUnique } from '../../typescript-tools'
import { DEMARCHES_TYPES_IDS, DemarcheTypeId } from '../demarchesTypes'
import { DocumentsTypes, DOCUMENTS_TYPES_IDS, EntrepriseDocumentTypeId, EntrepriseDocumentType, isEntrepriseDocumentTypeId } from '../documentsTypes'
import { ETAPES_TYPES, EtapeTypeId } from '../etapesTypes'
import { TitreTypeId, TITRES_TYPES_IDS } from '../titresTypes'
import { TDEType } from './index'

const EtapesTypesEntrepriseDocumentsTypes = {
  [ETAPES_TYPES.demande]: [
    DOCUMENTS_TYPES_IDS.attestationFiscale,
    DOCUMENTS_TYPES_IDS.avisDeSituationAuRepertoireSirene,
    DOCUMENTS_TYPES_IDS.curriculumVitae,
    DOCUMENTS_TYPES_IDS.identificationDeMateriel,
    DOCUMENTS_TYPES_IDS.justificatifDIdentite,
    DOCUMENTS_TYPES_IDS.justificatifDesCapacitesTechniques,
    DOCUMENTS_TYPES_IDS.kbis,
    DOCUMENTS_TYPES_IDS.justificatifDesCapacitesFinancieres,
  ],
  [ETAPES_TYPES.receptionDeComplements_RecepisseDeDeclarationLoiSurLeau_]: [DOCUMENTS_TYPES_IDS.kbis],
  [ETAPES_TYPES.modificationDeLaDemande]: [DOCUMENTS_TYPES_IDS.attestationFiscale, DOCUMENTS_TYPES_IDS.justificatifDesCapacitesFinancieres],
} as const satisfies { [key in EtapeTypeId]?: readonly EntrepriseDocumentTypeId[] }

const isEtapesTypesEntrepriseDocumentsTypes = (etapeTypeId?: EtapeTypeId | string): etapeTypeId is keyof typeof EtapesTypesEntrepriseDocumentsTypes => {
  return Object.keys(EtapesTypesEntrepriseDocumentsTypes).includes(etapeTypeId)
}

const TDEEntrepriseDocumentsTypes = {
  [TITRES_TYPES_IDS.AUTORISATION_DE_RECHERCHE_METAUX]: {
    [DEMARCHES_TYPES_IDS.Octroi]: {
      [ETAPES_TYPES.demande]: {
        [DOCUMENTS_TYPES_IDS.kbis]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.justificatifDesCapacitesFinancieres]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.justificatifDesCapacitesTechniques]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.justificatifDIdentite]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.attestationFiscale]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.curriculumVitae]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.identificationDeMateriel]: { optionnel: true },
      },
    },
  },
  [TITRES_TYPES_IDS.AUTORISATION_D_EXPLOITATION_METAUX]: {
    [DEMARCHES_TYPES_IDS.Octroi]: {
      [ETAPES_TYPES.demande]: {
        [DOCUMENTS_TYPES_IDS.justificatifDIdentite]: { optionnel: false, description: "Pièces d'identification du demandeur (personne physique ou société)" },
        [DOCUMENTS_TYPES_IDS.listeDesTravauxAnterieurs]: { optionnel: false, description: 'Liste des travaux auxquels le demandeur a participé au cours des 3 dernières années' },
        [DOCUMENTS_TYPES_IDS.justificatifDAdhesionALaCharteDesBonnesPratiques]: { optionnel: false },
        [DOCUMENTS_TYPES_IDS.attestationFiscale]: {
          optionnel: false,
          description:
            'Attestation sur l’honneur certifiant que le demandeur est en règle au regard de ses obligations fiscales et en matière de paiement de ses cotisations sociales (décret 2001-204, art. 7).',
        },
        [DOCUMENTS_TYPES_IDS.justificatifDesCapacitesTechniques]: {
          optionnel: false,
          description:
            'La justification du choix de la zone considérée, la situation du périmètre par rapport aux documents d’urbanisme, la situation du périmètre par rapport aux titres miniers et autorisations d’exploitation existants, les activités passées d’orpaillage et leurs conséquences sur les site.',
        },
        [DOCUMENTS_TYPES_IDS.TroisDerniersBilansEtComptesDeResultats]: {
          optionnel: false,
          description: 'Ces pièces sont demandées au titre de la justification des capacités financières du demandeur (décret 2001-204, art. 7)',
        },
        [DOCUMENTS_TYPES_IDS.referencesProfessionnelles]: {
          optionnel: false,
          description:
            'Références professionnelles du demandeur ou celles des cadres chargés du suivi et de la conduite des travauxCes pièces sont demandées au titre la justification des capacités techniques du demandeur (décret 2001-204, art. 6).\n \n S’il s’agit d’une personne morale, références professionnelles du ou des cadres chargés du suivi et de la conduite des travaux',
        },
        [DOCUMENTS_TYPES_IDS.declarationsBancairesOuCautionsAppropriees]: {
          optionnel: false,
          description: 'Ces pièces sont demandées au titre de la justification des capacités financières du demandeur (décret 2001-204, art. 7)',
        },
      },
    },
  },
} as const satisfies {
  [titreKey in TitreTypeId]?: {
    [demarcheKey in keyof TDEType[titreKey]]?: {
      [key in Extract<TDEType[titreKey][demarcheKey], readonly EtapeTypeId[]>[number]]?: { [key in EntrepriseDocumentTypeId]?: { optionnel: boolean; description?: string } }
    }
  }
}

type TDEEntrepriseDocumentsTypesUnleashed = {
  [key in TitreTypeId]?: { [key in DemarcheTypeId]?: { [key in EtapeTypeId]?: { [key in EntrepriseDocumentTypeId]: { optionnel: boolean; description?: string } } } }
}

export const getEntrepriseDocuments = (titreTypeId?: TitreTypeId, demarcheTypeId?: DemarcheTypeId, etapeTypeId?: EtapeTypeId): EntrepriseDocumentType[] => {
  if (isNotNullNorUndefined(titreTypeId) && isNotNullNorUndefined(demarcheTypeId) && isNotNullNorUndefined(etapeTypeId)) {
    const documentsIds: EntrepriseDocumentTypeId[] = []

    if (isEtapesTypesEntrepriseDocumentsTypes(etapeTypeId)) {
      documentsIds.push(...EtapesTypesEntrepriseDocumentsTypes[etapeTypeId])
    }

    documentsIds.push(...Object.keys((TDEEntrepriseDocumentsTypes as TDEEntrepriseDocumentsTypesUnleashed)[titreTypeId]?.[demarcheTypeId]?.[etapeTypeId] ?? {}).filter(isEntrepriseDocumentTypeId))

    return documentsIds.filter(onlyUnique).map(documentTypeId => {
      const documentSpecifique = (TDEEntrepriseDocumentsTypes as TDEEntrepriseDocumentsTypesUnleashed)[titreTypeId]?.[demarcheTypeId]?.[etapeTypeId]?.[documentTypeId]
      const document = { ...DocumentsTypes[documentTypeId], optionnel: true }
      if (documentSpecifique) {
        document.optionnel = documentSpecifique.optionnel
        document.description = documentSpecifique.description ?? document.description
      }

      return document
    })
  } else {
    throw new Error(`il manque des éléments pour trouver les documents titreTypeId: '${titreTypeId}', demarcheTypeId: ${demarcheTypeId}, etapeTypeId: ${etapeTypeId}`)
  }
}
