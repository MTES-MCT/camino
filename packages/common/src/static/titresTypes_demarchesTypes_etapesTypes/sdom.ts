import { DeepReadonly } from '../../typescript-tools.js'
import { DEMARCHES_TYPES_IDS } from '../demarchesTypes.js'
import { DOCUMENTS_TYPES_IDS } from '../documentsTypes.js'
import { DOMAINES_IDS } from '../domaines.js'
import { ETAPES_TYPES } from '../etapesTypes.js'
import { SDOMZoneId, SDOMZoneIds } from '../sdom.js'
import { toTitreTypeId } from '../titresTypes.js'
import { TITRES_TYPES_TYPES_IDS } from '../titresTypesTypes.js'

type DocumentTypeIdsBySdomZonesGet = 'nir' | 'jeg' | 'nip'
export const documentTypeIdsBySdomZonesGet = (sdomZones: DeepReadonly<SDOMZoneId[]>, titreTypeId: string, demarcheTypeId: string, etapeTypeId: string): DocumentTypeIdsBySdomZonesGet[] => {
  const documentTypeIds: DocumentTypeIdsBySdomZonesGet[] = []

  // Pour les demandes d’octroi d’AXM
  if (
    sdomZones.length > 0 &&
    etapeTypeId === ETAPES_TYPES.demande &&
    demarcheTypeId === DEMARCHES_TYPES_IDS.Octroi &&
    titreTypeId === toTitreTypeId(TITRES_TYPES_TYPES_IDS.AUTORISATION_D_EXPLOITATION, DOMAINES_IDS.METAUX)
  ) {
    if (sdomZones?.find(id => id === SDOMZoneIds.Zone2)) {
      // dans la zone 2 du SDOM les documents suivants sont obligatoires:
      documentTypeIds.push(DOCUMENTS_TYPES_IDS.noticeDImpactRenforcee)
      documentTypeIds.push(DOCUMENTS_TYPES_IDS.justificationDExistenceDuGisement)
    } else {
      documentTypeIds.push(DOCUMENTS_TYPES_IDS.noticeDImpact)
    }
  }

  return documentTypeIds
}
