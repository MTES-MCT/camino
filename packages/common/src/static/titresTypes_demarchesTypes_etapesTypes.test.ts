import { getDocuments } from './titresTypes_demarchesTypes_etapesTypes'
import { test, expect } from 'vitest'

test('getDocuments erreurs', () => {
  expect(() => getDocuments()).toThrowErrorMatchingInlineSnapshot(
    `"il manque des éléments pour trouver les documents domaineId: 'undefined', titreTypeType: undefined, demarcheId: undefined, etapeTypeId: undefined"`
  )
})

test('getDocuments pas de surcharge mais pas de documents', () => {
  expect(getDocuments('ap', 'm', 'amo', 'abs')).toMatchInlineSnapshot('[]')
})

test('getDocuments pas de surcharge', () => {
  expect(getDocuments('ap', 'm', 'amo', 'wfo')).toMatchInlineSnapshot(`
    [
      {
        "id": "dcl",
        "nom": "Déclaration",
        "optionnel": true,
      },
    ]
  `)
})

test('getDocuments surcharge', () => {
  expect(getDocuments('ax', 'm', 'oct', 'dae')).toMatchInlineSnapshot(`
    [
      {
        "description": undefined,
        "id": "arp",
        "nom": "Arrêté préfectoral",
        "optionnel": false,
      },
      {
        "id": "arr",
        "nom": "Arrêté",
        "optionnel": true,
      },
      {
        "id": "dep",
        "nom": "Décision cas par cas",
        "optionnel": true,
      },
      {
        "id": "let",
        "nom": "Lettre",
        "optionnel": true,
      },
      {
        "id": "ndd",
        "nom": "Notification de décision",
        "optionnel": true,
      },
    ]
  `)

  expect(getDocuments('ax', 'm', 'oct', 'mfr')).toMatchSnapshot()

  expect(getDocuments('ax', 'm', 'oct', 'dae')).not.toEqual(getDocuments('ax', 'm', 'ces', 'dae'))
})
