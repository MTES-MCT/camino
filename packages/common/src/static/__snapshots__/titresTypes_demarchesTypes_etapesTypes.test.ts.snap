// Vitest Snapshot v1

exports[`getDocuments surcharge 2`] = `
[
  {
    "id": "cam",
    "nom": "Contrat d'amodiation",
    "optionnel": true,
  },
  {
    "description": "Plan à l’échelle 1/50 000ème ou 1/100 000ème",
    "id": "car",
    "nom": "Documents cartographiques",
    "optionnel": false,
  },
  {
    "id": "cnt",
    "nom": "Contrat",
    "optionnel": true,
  },
  {
    "id": "cod",
    "nom": "Compléments au dossier de demande",
    "optionnel": true,
  },
  {
    "id": "dep",
    "nom": "Décision cas par cas",
    "optionnel": true,
  },
  {
    "id": "doe",
    "nom": "Dossier \\"Loi sur l'eau\\"",
    "optionnel": true,
  },
  {
    "id": "dom",
    "nom": "Dossier de demande",
    "optionnel": true,
  },
  {
    "id": "dos",
    "nom": "Dossier",
    "optionnel": true,
  },
  {
    "id": "fac",
    "nom": "Facture",
    "optionnel": true,
  },
  {
    "id": "fic",
    "nom": "Fiche de complétude",
    "optionnel": true,
  },
  {
    "id": "fip",
    "nom": "Fiche de présentation",
    "optionnel": true,
  },
  {
    "id": "for",
    "nom": "Formulaire de demande",
    "optionnel": true,
  },
  {
    "id": "jpa",
    "nom": "Justificatif de paiement",
    "optionnel": true,
  },
  {
    "id": "lem",
    "nom": "Lettre de demande",
    "optionnel": true,
  },
  {
    "description": undefined,
    "id": "let",
    "nom": "Lettre",
    "optionnel": true,
  },
  {
    "id": "noi",
    "nom": "Notice d'incidence",
    "optionnel": true,
  },
  {
    "id": "pla",
    "nom": "Plan",
    "optionnel": true,
  },
  {
    "id": "rec",
    "nom": "Récépissé \\"Loi sur l'eau\\"",
    "optionnel": true,
  },
  {
    "description": "la liste et la valeur du matériel d’extraction et de
 traitement que le demandeur détient ou qu’il 
envisage d’acquérir ainsi que, dans ce dernier
 cas, le financement correspondant. Ces pièces 
sont demandées au titre de la justification des 
capacités financières du
demandeur 
(décret 2001-204, art. 7)",
    "id": "idm",
    "nom": "Identification de matériel",
    "optionnel": false,
  },
  {
    "description": undefined,
    "id": "jeg",
    "nom": "Justification d’existence du gisement",
    "optionnel": true,
  },
  {
    "description": "la définition des mesures prévues par le pétitionnaire pour réhabiliter le site après exploitation, notamment la nature et les modalités de revégétalisation envisagée. (décret 2001-204, art. 5 bis)",
    "id": "mes",
    "nom": "Mesures prévues pour réhabiliter le site ",
    "optionnel": false,
  },
  {
    "description": "descriptif des méthodes envisagées pour l’exécution des travaux ((décret 2001-204, art. 6)",
    "id": "met",
    "nom": "Méthodes pour l'exécution des travaux",
    "optionnel": false,
  },
  {
    "description": undefined,
    "id": "nip",
    "nom": "Notice d’impact",
    "optionnel": true,
  },
  {
    "description": "Obligatoire pour les AEX hors de la zone 2 du SDOM",
    "id": "nir",
    "nom": "Notice d’impact renforcée",
    "optionnel": true,
  },
  {
    "description": "Description du phasage et planigramme des travaux. (décret 2001-204, art. 5)",
    "id": "prg",
    "nom": "Programme des travaux ",
    "optionnel": false,
  },
  {
    "description": "le schéma de pénétration du massif forestier proposé par le pétitionnaire pour l'acheminement du matériel lourd et la desserte du chantier (décret 2001-204, art. 5 bis)",
    "id": "sch",
    "nom": "Schéma de pénétration du massif forestier",
    "optionnel": false,
  },
]
`;
